### ISSUE TITLE
Feature request: Parallel computing in command substitution

### ISSUE THREAD

**Issue by sulivanShu** [fish-shell/fish-shell#8038]
fish, version 3.2.2

Hi,

To my knowledge, Fish does not offer any support for parallelization of subcommands.

Bash offers partial support, only for sub-processes.

It might be a nice feature to have some support for parallelization of subcommands, all subcommands being executed in parallel.

Some benchmark between Bash and Fish:

```
~> time fish -c 'diff -u (sleep 1 ; echo hello | psub) (sleep 1 ; echo hella | psub)' # parallelization: no
--- /tmp/.psub.F1JEGKhmAv       2021-05-31 23:23:01.519767985 +0200
+++ /tmp/.psub.vt6aNxURbp       2021-05-31 23:23:02.539763002 +0200
@@ -1 +1 @@
-hello
+hella

________________________________________________________
Executed in    2,11 secs      fish           external
   usr time   97,80 millis    3,49 millis   94,30 millis
   sys time   20,83 millis    0,00 millis   20,83 millis

~> time bash -c 'diff -u <(sleep 1 ; echo hello) <(sleep 1 ; echo hella)' # parallelization: yes
--- /dev/fd/63  2021-05-31 23:23:23.149662937 +0200
+++ /dev/fd/62  2021-05-31 23:23:23.149662937 +0200
@@ -1 +1 @@
-hello
+hella

________________________________________________________
Executed in    1,05 secs      fish           external
   usr time   19,62 millis    2,81 millis   16,81 millis
   sys time    1,17 millis    1,17 millis    0,00 millis

~> time fish -c 'echo (sleep 1; echo hello) (sleep 1; echo hello)' # parallelization: no
hello hello

________________________________________________________
Executed in    2,05 secs      fish           external
   usr time   59,05 millis    1,78 millis   57,28 millis
   sys time    3,50 millis    0,74 millis    2,75 millis

~> time bash -c 'echo $(sleep 1; echo hello) $(sleep 1; echo hello)' # parallelization: no
hello hello

________________________________________________________
Executed in    2,04 secs      fish           external
   usr time   12,90 millis    3,00 millis    9,90 millis
   sys time   24,63 millis    1,26 millis   23,37 millis
```

**Comment by faho** [comment]
Duplicate of #1439 or #1396 - the way bash does this is via subshells, which involves a separate environment and means you can't just set a variable here and expect it to be usable outside.

**Comment by sulivanShu** [comment]
I found an alternative:

```
mkfifo {(seq --separator=, 10)} ;
sh -c "sleep 1 ; echo A > 1" & # long command in parallel
sh -c "sleep 2 ; echo B > 2" & # other long command
sh -c "sleep 1 ; echo C > 3" & 
sh -c "sleep 3 ; echo D > 4" & 
sh -c "sleep 4 ; echo E | tee 4 5 > /dev/null" & # Warning! tee is serial: 5 after 4
# sh -c "sleep 5 ; echo F | tee 6 7 > /dev/null" &
cat 1 2 & # subprocesses equivalent
echo (cat 3)\n(cat 4) & # works in subcommands
cat 4 &
cat 5 & # 4 and 5 in separate commands is ok
# cat 5 & # Freezing. Cannot reuse fifo. Use temp files instead.
# cat 6 7 & # Freezing. In input, cat waits for 7 begins, but in output, tee waits for 6 ends. Use temp files. 
wait ;
rm {(seq --separator=, 10)}
```
PS: Is this the right place to chat? Or I have to publish my messages in #1439 or #1396 ?
